### ISSUE TITLE
Promise should be optional

### ISSUE THREAD

**Issue by JSteunou** [centrifugal/centrifuge-js#43]
es6-promise is not the heaviest but still, I could save some KB if centrifuge made it optional like SockJS. In our modern world, considering only not deprecated / unsupported browsers, and nodejs > LTS, Promise is nearly everywhere in native.

http://caniuse.com/#feat=promises and nodejs >= 0.12

**Comment by FZambia** [comment]
@JSteunou hello! This is reasonable, do you know a way to do this in backwards compatible manner?

**Comment by JSteunou** [comment]
If you want to not introduce breaking changes you will have to make 2 builds. Both would have a Promise option like you did for SockJS then one would have a default value set to es6-promise, that will be the backward compatible version, and the other version would have this option set to global Promise.
On the next major release you could drop the es6-promise version

**Comment by FZambia** [comment]
Thanks! Yep, that is the only way I supposed... Managing 2 builds looks like a pain and mess - and it seems that result won't be noticable by most users as es6-promise is only 2.4kb gzipped.

So let's put it this way - we will make Promise optional in next major release when it comes. Is it ok? Can you live with it now or in practice that extra kilobytes make real difference for your case?

**Comment by JSteunou** [comment]
I can live with it 😊

If you need some help for the next major release I could make some PR. Mostly around the build, using rollup to produce clear dist files for all targets (node, browser legacy modern...)

**Comment by Inpassor** [comment]
@JSteunou have a look at #53 
Maybe you will find it helpful.

**Comment by JSteunou** [comment]
It might, but I do not use .ts in my projects.

**Comment by Inpassor** [comment]
@JSteunou so do not use it.
`import {Centrifuge} from 'centrifuge' ` should import  **./dist/centrifuge.js**

**Comment by JSteunou** [comment]
Ho. Isnt that weird?

**Comment by JSteunou** [comment]
My bad I did not understood the all thing, I though it was a PR to add the ts verison in *this* project, instead of another project with another name. Will try that.

**Comment by Inpassor** [comment]
@JSteunou correct. This is the another project [https://github.com/Inpassor/centrifuge-ts](url). NPM package have name **centrifuge-ts**.

**Comment by FZambia** [comment]
@JSteunou while working on Centrifugo v2 and Centrifuge library this client was adapted to work with new release. You can look at its current state in [c2](https://github.com/centrifugal/centrifuge-js/tree/c2) branch. 

While investigating which browsers new client should support out of the box I found that it seems reasonable to support IE 11 these days which has no Promise so library users must include polyfill anyway. So at moment Promise dependency still in client - even in v2. I believe that having support for IE 11 (~3% global use) out of the box can simplify life for developers - as far as I understand most devs still want to support that IE version. Promise is a core client dependency - it's not very similar to SockJS which can be completely omitted by users.

**Comment by JSteunou** [comment]
Sorry but I still disagree. I support IE11 too but as you said, this is a very low usage, so

1. it is better and it is a good practice to target your user agent and serve the needed polyfill instead of wrapping it up in the final app.
2. This is a library, not even a final app. If all library authors do include all the needed polyfill, our app size would scratch the sky!! Can you believe it?!

My 2 cents: produce 2 builds, one with, one without the polyfill. It will make everyone happy :)

**Comment by FZambia** [comment]
Unfortunately in version 2 we already have two different builds - one with Protobuf and one without:) So splitting builds more is no choice here.

**Comment by JSteunou** [comment]
Then you might follow the 3 years ago started march and remove Promise polyfill in favor of a readme IE11 section with an example like 

```
<!-- Script for polyfilling Promises -->
<script src='https://cdn.polyfill.io/v2/polyfill.min.js'></script>
```
You might also read https://philipwalton.com/articles/loading-polyfills-only-when-needed/ 

We are in 2018, this should be obvious that libraries must not embed polyfill. I think you're the last one doing it :(

**Comment by JSteunou** [comment]
Sure thing. Thanks you for hearing me out ;)

**Comment by JSteunou** [comment]
@FZambia your c2 branch looks fine, I just have a slight preference for rollup over webpack to build libraries like this one. I like to use rollup + babel 7 and put those main fields definition in package.json

```json
{
"browser": "dist/lib.min.js",
"main": "dist/lib.js",
"module": "src/index.js or dist/lib.esm.js"
}
```

**Comment by JSteunou** [comment]
I can help with that if you want. Do you plan to release this v2 soon?

**Comment by JSteunou** [comment]
Webpack or Rollup is sometimes just a personal matter, but when it comes to application bundle versus libraries bundle, Rollup is often prefered over Webpack for libraries. This is because Rollup comes with minimal configuration (but since Webpack 4 you can also be with less configuration) and has a lighter overhead on final bundle.

I find it also easier to produce `umd` & `esm` bundle with it. Exposing named export in `esm` bundle.

I can update the Webpack config to use babel 7 and make a separate branch aside with the equivalent with Rollup so you can compare. From which branch should I start?

**Comment by FZambia** [comment]
Still advantages are not obvious for me. Also we need to keep the following things working:

1) Separate builds for JSON and Protobuf with tree shaking. Current output of `npm run build` with latest source is:

```
                         Asset     Size  Chunks             Chunk Names
    centrifuge.protobuf.min.js   105 kB       0  [emitted]  centrifuge.protobuf
             centrifuge.min.js  34.5 kB       1  [emitted]  centrifuge
centrifuge.protobuf.min.js.map   785 kB       0  [emitted]  centrifuge.protobuf
         centrifuge.min.js.map   174 kB       1  [emitted]  centrifuge
``` 

2) Work everywhere - browser and Node.js
3) Have the same file names in `dist` folders so users could import libraries like:

```
<script src="//cdn.rawgit.com/centrifugal/centrifuge-js/2.X.X/dist/centrifuge.min.js"></script>
```

And for Protobuf version:

```
<script src="//cdn.rawgit.com/centrifugal/centrifuge-js/2.X.X/dist/centrifuge.protobuf.min.js"></script>
```

4) When developing with `npm` users should have a possibility to import JSON version of `Centrifuge` object by default and do something like this:

```
import Centrifuge from 'centrifuge/dist/centrifuge.protobuf';
```

If they need Protobuf version. This is the only approach I found to keep tree shaking working when importing library installled via `npm` from application source code (I tried several approaches but current way is the only one that worked fine for me).


Can we keep all these properties with rollup?

**Comment by JSteunou** [comment]
Yup, sorry I did not find time yet to make a PR to show you how to use it, but still on my todo list ;)
